File: app/src/test/java/org/thoughtcrime/securesms/testutil/SystemOutLogger.java
package org.thoughtcrime.securesms.testutil;

import org.signal.core.util.logging.Log;

public final class SystemOutLogger extends Log.Logger {
  @Override
  public void v(String tag, String message, Throwable t, boolean keepLonger) {
    printlnFormatted('v', tag, message, t);
  }

  @Override
  public void d(String tag, String message, Throwable t, boolean keepLonger) {
    printlnFormatted('d', tag, message, t);
  }

  @Override
  public void i(String tag, String message, Throwable t, boolean keepLonger) {
    printlnFormatted('i', tag, message, t);
  }

  @Override
  public void w(String tag, String message, Throwable t, boolean keepLonger) {
    printlnFormatted('w', tag, message, t);
  }

  @Override
  public void e(String tag, String message, Throwable t, boolean keepLonger) {
    printlnFormatted('e', tag, message, t);
  }

  @Override
  public void flush() { }

  private void printlnFormatted(char level, String tag, String message, Throwable t) {
    System.out.println(format(level, tag, message, t));
  }

  private String format(char level, String tag, String message, Throwable t) {
    if (t != null) {
      t.printStackTrace();
      return String.format("%c[%s] %s %s:%s", level, tag, message, t.getClass().getSimpleName(), t.getMessage());
    } else {
      return String.format("%c[%s] %s", level, tag, message);
    }
  }
}

File: feature/registration/src/test/java/org/signal/registration/fakes/SystemOutLogger.kt
/*
 * Copyright 2026 Signal Messenger, LLC
 * SPDX-License-Identifier: AGPL-3.0-only
 */

package org.signal.registration.fakes

import org.signal.core.util.logging.Log

/**
 * Writes logs to stdout so they show up in unit test output.
 */
class SystemOutLogger : Log.Logger() {
  override fun v(tag: String, message: String?, t: Throwable?, keepLonger: Boolean) = print("V", tag, message, t)
  override fun d(tag: String, message: String?, t: Throwable?, keepLonger: Boolean) = print("D", tag, message, t)
  override fun i(tag: String, message: String?, t: Throwable?, keepLonger: Boolean) = print("I", tag, message, t)
  override fun w(tag: String, message: String?, t: Throwable?, keepLonger: Boolean) = print("W", tag, message, t)
  override fun e(tag: String, message: String?, t: Throwable?, keepLonger: Boolean) = print("E", tag, message, t)
  override fun flush() = Unit

  private fun print(level: String, tag: String, message: String?, t: Throwable?) {
    println("[$level][$tag] $message")
    t?.printStackTrace(System.out)
  }
}
